### PR TITLE
Implement assignment operator for common::optional.

### PR DESCRIPTION
**Pull request by spielawa**
Implement assignment operator for common::optional.

### CODE REVIEW

File: cartographer/common/optional.h [hunk 1]
```diff
@@ -45,6 +45,20 @@ class optional {
     return *value_;
   }
 
+  optional<T> operator=(const T& other_value) {
```
Comment (wohe) [inline cartographer/common/optional.h]: The return value should be of type `optional<T>&`. Similar below.

File: cartographer/common/optional.h [hunk 2]
```diff
@@ -45,6 +45,20 @@ class optional {
...
+    this->value_ = common::make_unique<T>(other_value);
+    return *this;
+  }
+
+  optional<T> operator=(const optional<T>& other) {
+    if (!other.has_value()) {
+      this->value_ = nullptr;
```
Comment (wohe) [inline cartographer/common/optional.h]: Add a test for this behavior?

**Comment by wohe** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user wohe. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Error: exit status 128

Log:
~~~
=> handle work item: PR 800 on googlecartographer/cartographer.
 => Switching to branch master.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at 8165da8 Add GetConstraints() to gRPC service. (#798)
 $ git clean -fd [data/cartographer]
 $ git checkout master [data/cartographer]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/cartographer]
Already up to date.
Current branch master is up to date.
 => Checking out spielawa/overload_operator_in_optional as local branch pr_800.
 $ git remote add spielawa git@github.com:spielawa/cartographer.git [data/cartographer]
fatal: remote spielawa already exists.
 $ git fetch spielawa [data/cartographer]
Received disconnect from 192.30.253.113 port 22:11: Bye Bye
Disconnected from 192.30.253.113 port 22
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

~~~

**Comment by SirVer** [comment]
No idea what failed here. Maybe a transient error. Let's try again.

@wally-the-cartographer merge

**Comment by cschuet** [comment]
@SirVer Yes looks like the Github server was not feeling well.

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 1.

**Pull request merged**